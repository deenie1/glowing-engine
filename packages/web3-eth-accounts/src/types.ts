import { FeeMarketEIP1559TxData, AccessListEIP2930TxData, TxData } from '@ethereumjs/tx';
import { HexString } from 'web3-utils';

export type signatureObject = {
	messageHash: string;
	r: string;
	s: string;
	v: string;
};

export const keyStoreSchema = {
	type: 'object',
	required: ['crypto', 'id', 'version', 'address'],
	properties: {
		crypto: {
			type: 'object',
			required: ['cipher', 'ciphertext', 'cipherparams', 'kdf', 'kdfparams', 'mac'],
			properties: {
				cipher: { type: 'string' },
				ciphertext: { type: 'string' },
				cipherparams: { type: 'object' },
				kdf: { type: 'string' },
				kdfparams: { type: 'object' },
				salt: { type: 'string' },
				mac: { type: 'string' },
			},
		},
		id: { type: 'string' },
		version: { type: 'number' },
		address: { type: 'string' },
	},
};

export type signTransactionResult = signatureObject & {
	rawTransaction: string;
	transactionHash: string;
};

export type signTransactionFunction = (
	transaction: TxData | AccessListEIP2930TxData | FeeMarketEIP1559TxData,
	privateKey: string,
) => signTransactionResult;

export type signResult = signatureObject & {
	message?: string;
	signature: string;
};

export type signFunction = (data: string, privateKey: string) => signResult;

// https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition

export type Cipher = 'aes-128-ctr' | 'aes-128-cbc' | 'aes-256-cbc';

export type CipherOptions = {
	salt?: Buffer | string;
	iv?: Buffer | string;
	kdf?: 'scrypt' | 'pbkdf2';
	dklen?: number;
	c?: number; // iterrations
	n?: number; // cpu/memory cost
	r?: number; // block size
	p?: number; // parallelization cost
};

export type ScryptParams = {
	dklen: number;
	n: number;
	p: number;
	r: number;
	salt: Buffer | string;
};
export type PBKDF2SHA256Params = {
	c: number; // iterations
	dklen: number;
	prf: 'hmac-sha256';
	salt: Buffer | string;
};

export type KeyStore = {
	crypto: {
		cipher: Cipher;
		ciphertext: string;
		cipherparams: {
			iv: string;
		};
		kdf: 'pbkdf2' | 'scrypt';
		kdfparams: ScryptParams | PBKDF2SHA256Params;
		mac: HexString;
	};
	id: string;
	version: 3;
	address: string;
};
