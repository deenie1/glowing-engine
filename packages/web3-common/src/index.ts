export * from './errors';
export * from './constants';
export * from './types';
export * from './web3_base_provider';
export * from './web3_base_wallet';
export * from './web3_event_emitter';
export * from './eth_execution_api';
export * from './deferred_promise';
export * as jsonRpc from './json_rpc';
