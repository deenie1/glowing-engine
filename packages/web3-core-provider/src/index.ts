import {
    Web3Client,
    Eip1193Provider,
    IWeb3Provider,
} from 'web3-core-types/lib/types';
import Web3ProvidersHttp from 'web3-providers-http';
import Web3ProvidersEip1193 from 'web3-providers-eip1193';
import Web3ProviderWS from 'web3-providers-ws';

import { ClientProtocol } from './types';

/**
 * Detects protocol for provided {web3Client}, and instantiates
 * corresponding provider package
 *
 * @param web3Client Client passed to provider package for instantiation
 * @returns void
 */
export default function initWeb3Provider(
    web3Client: Web3Client
): IWeb3Provider {
    switch (detectClientProtocol(web3Client)) {
        case ClientProtocol.Eip1193:
            return new Web3ProvidersEip1193(web3Client as Eip1193Provider);
        case ClientProtocol.HTTP:
            return new Web3ProvidersHttp(web3Client as string);
        case ClientProtocol.WS:
            return new Web3ProviderWS(web3Client as string);
        case ClientProtocol.IPC:
            // TODO
            throw Error('Provider protocol not implemented');
        default:
            throw Error('Provider protocol not supported');
    }
}

/**
 * Detects protocol of provided client
 *
 * @param web3Client Client to be used for provider instantiation
 * @returns ClientProtocol enum value
 */
function detectClientProtocol(web3Client: Web3Client): ClientProtocol {
    try {
        if (
            typeof web3Client === 'object' &&
            web3Client.request !== undefined
        ) {
            return ClientProtocol.Eip1193;
        } else if (
            typeof web3Client === 'string' &&
            /^http(s)?:\/\//i.test(web3Client)
        ) {
            return ClientProtocol.HTTP;
        } else if (
            typeof web3Client === 'string' &&
            /^ws(s)?:\/\//i.test(web3Client)
        ) {
            return ClientProtocol.WS;
        } else if (
            typeof web3Client === 'string' &&
            /^ipc:\/\//i.test(web3Client)
        ) {
            return ClientProtocol.IPC;
        }
        return ClientProtocol.UNKNOWN;
    } catch (error) {
        throw Error(`Error detecting client protocol: ${error.message}`);
    }
}
