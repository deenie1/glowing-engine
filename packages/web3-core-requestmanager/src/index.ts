import Web3ProviderHttp from 'web3-providers-http';
import {
    ProviderOptions,
    ProviderCallOptions,
    RpcOptions,
    RpcResponse,
    PartialRpcOptions,
    SubscriptionResponse,
    CallOptions,
} from 'web3-providers-base/src/types';

import { ProviderProtocol } from './types';

export default class Web3RequestManager {
    provider: Web3ProviderHttp | undefined;

    providerProtocol: ProviderProtocol = ProviderProtocol.UNKNOWN;

    private static _DEFAULT_JSON_RPC_VERSION = '2.0';

    constructor(providerOptions: ProviderOptions) {
        switch (
            Web3RequestManager.detectProviderProtocol(
                providerOptions.providerUrl
            )
        ) {
            case ProviderProtocol.HTTP:
                this.provider = new Web3ProviderHttp(providerOptions);
                this.providerProtocol = ProviderProtocol.HTTP;
                break;
            case ProviderProtocol.WS:
                this.providerProtocol = ProviderProtocol.WS;
                // TODO
                throw Error('Provider protocol not implemented');
            case ProviderProtocol.IPC:
                this.providerProtocol = ProviderProtocol.IPC;
                // TODO
                throw Error('Provider protocol not implemented');
            default:
                // TODO figure out if possible to support generic provider
                throw Error('Provider protocol not supported');
        }
    }

    static detectProviderProtocol(providerUrl: string): ProviderProtocol {
        try {
            if (/^http(s)?:\/\//i.test(providerUrl)) {
                return ProviderProtocol.HTTP;
            } else if (/^ws(s)?:\/\//i.test(providerUrl)) {
                return ProviderProtocol.WS;
            } else if (/^ipc:\/\//i.test(providerUrl)) {
                return ProviderProtocol.WS;
            }
            return ProviderProtocol.UNKNOWN;
        } catch (error) {
            throw Error(`Error detecting provider protocol: ${error.message}`);
        }
    }

    private static _defaultRpcOptions(
        rpcOptions: PartialRpcOptions
    ): RpcOptions {
        return {
            id:
                rpcOptions.id ||
                Math.floor(Math.random() * Number.MAX_SAFE_INTEGER), // generate random integer
            jsonrpc: rpcOptions.jsonrpc || this._DEFAULT_JSON_RPC_VERSION,
            method: rpcOptions.method,
            params: rpcOptions.params,
        };
    }

    async send(callOptions: CallOptions): Promise<RpcResponse> {
        try {
            if (this.provider === undefined)
                throw Error('No provider initialized');
            switch (this.providerProtocol) {
                case ProviderProtocol.WS:
                    // TODO
                    throw Error('Provider protocol not implemented');
                case ProviderProtocol.IPC:
                    // TODO
                    throw Error('Provider protocol not implemented');
                default:
                    // this.providerProtocol is assumed to be ProviderProtocol.HTTP
                    // return callOptions.rpcOptions
                    //     ? this.provider.send(
                    //           //Checks if RPC options exist
                    //           callOptions.providerCallOptions,
                    //           Web3RequestManager._defaultRpcOptions(callOptions.rpcOptions)
                    //       )
                    //     : this.provider.send(callOptions.providerCallOptions);

                    const defaultedCallOptions: CallOptions = {
                        ...callOptions,
                        rpcOptions: callOptions.rpcOptions
                            ? Web3RequestManager._defaultRpcOptions(
                                  callOptions.rpcOptions
                              )
                            : undefined,
                    };

                    return this.provider.send(defaultedCallOptions);
            }
        } catch (error) {
            throw Error(`Error sending: ${error.message}`);
        }
    }

    async subscribe(
        providerCallOptions: ProviderCallOptions,
        rpcOptions?: PartialRpcOptions
    ): Promise<SubscriptionResponse> {
        try {
            if (this.provider === undefined)
                throw Error('No provider initialized');
            return rpcOptions
                ? this.provider.subscribe(
                      providerCallOptions,
                      Web3RequestManager._defaultRpcOptions(rpcOptions)
                  )
                : this.provider.subscribe(providerCallOptions);
        } catch (error) {
            throw Error(`Error subscribing: ${error.message}`);
        }
    }
}
