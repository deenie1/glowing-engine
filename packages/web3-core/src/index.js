/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file index.js
 * @author Fabian Vogelsteller <fabian@ethereum.org>
 * @date 2017
 */

const { Manager, BatchManager } = require("web3-core-requestmanager");
const extend = require("./extend");






class Core {
  constructor (...args) {
    // sets _requestmanager etc
    if (args[0] && args[0]._requestManager) {
        this._requestManager = args[0]._requestManager;
    } else {
        this._requestManager = new Manager(args[0], args[1]);
    }

    // add givenProvider
    this.givenProvider = Manager.givenProvider;
    this.providers = Manager.providers;

    this._provider = this._requestManager.provider;

    // attach batch request creation
    this.BatchRequest = BatchManager.bind(null, this._requestManager);

    // attach extend function
    this.extend = extend(this);
  }

  setRequestManager (manager) {
    this._requestManager = manager;
    this._provider = manager.provider;
  }

  set curentProvider (newProvider) {
    return this.setProvider(newProvider);
  }

  get curentProvider () {
    return this._provider;
  }

  static get providers () {
    return Manager.providers
  }

  static get givenProvider () {
    return Manager.givenProvider
  }

  setProvider (provider, net) {
    this._requestManager.setProvider(provider, net);
    this._provider = this._requestManager.provider;
    this.extend = extend(this);
    return true;
  }

  addProviders () {
    this.givenProvider = Manager.givenProvider;
    this.providers = Manager.providers;
  }

  static addProviders (pkg) {
    pkg.givenProvider = Manager.givenProvider;
    pkg.providers = Manager.providers;
  }

  static packageInit (pkg, args) {
    args = Array.prototype.slice.call(args);

    if (!pkg) {
        throw new Error('You need to instantiate using the "new" keyword.');
    }

    // make property of pkg._provider, which can properly set providers
    Object.defineProperty(pkg, 'currentProvider', {
        get: () => {
            return pkg._provider;
        },
        set: (value) => {
            return pkg.setProvider(value);
        },
        enumerable: true,
        configurable: true
    });

    // inherit from parent package or create a new RequestManager
    if (args[0] && args[0]._requestManager) {
        pkg._requestManager = args[0]._requestManager;
    } else {
        pkg._requestManager = new Manager(args[0], args[1]);
    }

    // add givenProvider
    pkg.givenProvider = Manager.givenProvider;
    pkg.providers = Manager.providers;

    pkg._provider = pkg._requestManager.provider;

    // add SETPROVIDER function (don't overwrite if already existing)
    if (!pkg.setProvider) {
        pkg.setProvider = (provider, net) => {
            pkg._requestManager.setProvider(provider, net);
            pkg._provider = pkg._requestManager.provider;
            return true;
        };
    }

    pkg.setRequestManager = (manager) => {
        pkg._requestManager = manager;
        pkg._provider = manager.provider;
    };

    // attach batch request creation
    pkg.BatchRequest = BatchManager.bind(null, pkg._requestManager);

    // attach extend function
    pkg.extend = extend(pkg);
  }



}

module.exports = Core
