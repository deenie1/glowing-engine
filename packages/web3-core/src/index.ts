/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/

export * from './web3_config';
export * from './web3_request_manager';
export * from './web3_subscription_manager';
export * from './web3_subscriptions';
export * from './web3_context';
export * from './web3_batch_request';
export * from './utils';
export * from './types';
export * from './formatters';
export * from './web3_promi_event';
export * from './web3_event_emitter';

// For backward usability export as namespace
export * as formatters from './formatters';
