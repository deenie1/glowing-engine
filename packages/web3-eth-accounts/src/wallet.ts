import {
	Web3BaseWallet,
	Web3BaseWalletAccount,
	Web3AccountProvider,
	Web3EncryptedWallet,
} from 'web3-common';

type BrowserError = { code: number; name: string };

export class Wallet<
	T extends Web3BaseWalletAccount = Web3BaseWalletAccount,
> extends Web3BaseWallet<T> {
	private readonly _accounts: { [key: string]: T };
	private readonly _defaultKeyName = 'web3js_wallet';

	public constructor(accountProvider: Web3AccountProvider<T>) {
		super(accountProvider);
		this._accounts = {};
	}

	public static getStorage(): Storage | undefined {
		let storage: Storage | undefined;

		try {
			storage = window.localStorage;
			const x = '__storage_test__';
			storage.setItem(x, x);
			storage.removeItem(x);

			return storage;
		} catch (e: unknown) {
			return (e as BrowserError) &&
				// everything except Firefox
				((e as BrowserError).code === 22 ||
					// Firefox
					(e as BrowserError).code === 1014 ||
					// test name field too, because code might not be present
					// everything except Firefox
					(e as BrowserError).name === 'QuotaExceededError' ||
					// Firefox
					(e as BrowserError).name === 'NS_ERROR_DOM_QUOTA_REACHED') &&
				// acknowledge QuotaExceededError only if there's something already stored
				storage !== undefined &&
				storage.length !== 0
				? storage
				: undefined;
		}
	}

	public get length() {
		return Object.keys(this._accounts).length;
	}

	public create(numberOfAccounts: number) {
		for (let i = 0; i < numberOfAccounts; i += 1) {
			this.add(this._accountProvider.create());
		}

		return this;
	}

	public add(account: T | string): boolean {
		if (typeof account === 'string') {
			return this.add(this._accountProvider.privateKeyToAccount(account));
		}

		this._accounts[account.address.toLowerCase()] = account;

		return true;
	}

	public get(addressOrIndex: string | number): T {
		if (typeof addressOrIndex === 'string') {
			return this._accounts[addressOrIndex];
		}

		return Object.values(this._accounts)[addressOrIndex];
	}

	public remove(addressOrIndex: string | number): boolean {
		const result =
			typeof addressOrIndex === 'string'
				? { address: addressOrIndex }
				: Object.values(this._accounts)[addressOrIndex];

		if (result && this._accounts[result.address]) {
			delete this._accounts[result.address];
			return true;
		}

		return false;
	}

	public clear() {
		for (const key of Object.keys(this._accounts)) {
			delete this._accounts[key];
		}

		return this;
	}

	public encrypt(password: string, options?: Record<string, unknown> | undefined) {
		return Object.values(this._accounts).map(account => account.encrypt(password, options));
	}

	public decrypt(
		encryptedWallets: string[],
		password: string,
		options?: Record<string, unknown> | undefined,
	) {
		for (const wallet of encryptedWallets) {
			this.add(this._accountProvider.decrypt(wallet, password, options));
		}

		return this;
	}

	public save(password: string, keyName?: string) {
		const storage = Wallet.getStorage();

		if (!storage) {
			throw new Error('Local storage not available.');
		}

		storage.setItem(keyName ?? this._defaultKeyName, JSON.stringify(this.encrypt(password)));

		return true;
	}

	public load(password: string, keyName?: string) {
		const storage = Wallet.getStorage();

		if (!storage) {
			throw new Error('Local storage not available.');
		}

		const keystore = storage.getItem(keyName ?? this._defaultKeyName);

		if (keystore) {
			this.decrypt((JSON.parse(keystore) as Web3EncryptedWallet[]) || [], password);
		}

		return this;
	}
}
