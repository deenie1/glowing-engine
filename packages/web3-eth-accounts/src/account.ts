import { utils, getPublicKey } from 'ethereum-cryptography/secp256k1';
import {
	toChecksumAddress,
	bytesToHex,
	sha3Raw,
	HexString,
	isBuffer,
	isValidString,
} from 'web3-utils';
import { InvalidPrivateKeyError, PrivateKeyLengthError } from 'web3-common';

// TODO Will be added later
export const encrypt = (): boolean => true;

// TODO Will be added later
export const sign = (): boolean => true;

// TODO Will be added later
export const signTransaction = (): boolean => true;

/**
 * Get account from private key
 */
export const privateKeyToAccount = (
	privateKey: string | Buffer,
): {
	address: string;
	privateKey: string;
	signTransaction: () => boolean; // From 1.x
	sign: () => boolean;
	encrypt: () => boolean;
} => {
	if (!(isValidString(privateKey) || isBuffer(privateKey))) {
		throw new InvalidPrivateKeyError(privateKey);
	}

	const stringPrivateKey = Buffer.isBuffer(privateKey)
		? Buffer.from(privateKey).toString('hex')
		: privateKey;

	const stringPrivateKeyNoPrefix = stringPrivateKey.startsWith('0x')
		? stringPrivateKey.slice(2)
		: stringPrivateKey;

	// TODO Replace with isHexString32Bytes function in web3-eth PR:
	// Must be 64 hex characters
	if (stringPrivateKeyNoPrefix.length !== 64) {
		throw new PrivateKeyLengthError(stringPrivateKeyNoPrefix);
	}

	const publicKey = getPublicKey(stringPrivateKeyNoPrefix);

	const publicKeyString = `0x${publicKey.slice(2)}`;
	const publicHash = sha3Raw(publicKeyString);
	const publicHashHex = bytesToHex(publicHash);
	const address = toChecksumAddress(publicHashHex.slice(-40)); // To get the address, take the last 20 bytes of the public hash
	return { address, privateKey: stringPrivateKey, signTransaction, sign, encrypt };
};

/**
 * Returns an acoount
 */
export const create = (): {
	address: HexString;
	privateKey: string;
	signTransaction: () => boolean; // From 1.x
	sign: () => boolean;
	encrypt: () => boolean;
} => {
	const privateKey = utils.randomPrivateKey();
	const address = getPublicKey(privateKey);
	return {
		privateKey: `0x${Buffer.from(privateKey).toString('hex')}`,
		address: `0x${Buffer.from(address).toString('hex')}`,
		signTransaction,
		sign,
		encrypt,
	};
};
