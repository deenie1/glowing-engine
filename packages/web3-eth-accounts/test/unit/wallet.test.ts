/* eslint-disable @typescript-eslint/no-magic-numbers */

import { when } from 'jest-when';
import { Web3AccountProvider, Web3BaseWalletAccount } from 'web3-common';
import { Wallet } from '../../src/wallet';

describe('Wallet', () => {
	let wallet: Wallet;
	let accountProvider: Web3AccountProvider<Web3BaseWalletAccount>;
	let localStorageSpy: {
		getItem: jest.MockedFunction<(key: string) => string>;
		setItem: jest.MockedFunction<(key: string, value: string) => string>;
	};
	let totalAccountsCreate = 0;
	let totalPrivatekeyCreate = 0;

	beforeEach(() => {
		localStorageSpy = { getItem: jest.fn(), setItem: jest.fn() };

		jest.spyOn(Wallet, 'getStorage').mockReturnValue(localStorageSpy as never);

		accountProvider = {
			privateKeyToAccount: jest.fn().mockImplementation(() => {
				totalPrivatekeyCreate += 1;
				return { address: `privatekey_create_${totalPrivatekeyCreate}` };
			}),
			decrypt: jest.fn(),
			create: jest.fn().mockImplementation(() => {
				totalAccountsCreate += 1;
				return { address: `account_create_${totalAccountsCreate}` };
			}),
		};
		wallet = new Wallet(accountProvider);
	});

	describe('constructor', () => {
		it('should create instance of the wallet', () => {
			expect(wallet).toBeInstanceOf(Wallet);
		});
	});

	describe('create', () => {
		it('should create given # of accounts using the account provider', () => {
			const numberOfAccounts = 10;
			const accounts = [];

			wallet.create(numberOfAccounts);

			for (let i = 0; i < numberOfAccounts; i += 1) {
				accounts.push(wallet.get(i));
			}

			expect(wallet).toHaveLength(numberOfAccounts);
			expect(accounts).toHaveLength(numberOfAccounts);
			expect(accounts).toMatchSnapshot();
			expect(accountProvider.create).toHaveBeenCalledTimes(numberOfAccounts);
		});
	});

	describe('add', () => {
		it('should create account from private key if string value is given', () => {
			const privateKey = 'private key';
			const result = wallet.add(privateKey);

			expect(result).toBeTruthy();

			expect(accountProvider.privateKeyToAccount).toHaveBeenCalledTimes(1);
			expect(accountProvider.privateKeyToAccount).toHaveBeenCalledWith(privateKey);
			expect(wallet).toHaveLength(1);
			expect(wallet.get(0)).toEqual({ address: 'privatekey_create_1' });
		});

		it('should not create account from private key if object value is given', () => {
			const result = wallet.add({ address: 'my_address' } as never);

			expect(result).toBeTruthy();
			expect(accountProvider.privateKeyToAccount).toHaveBeenCalledTimes(0);
			expect(wallet).toHaveLength(1);
			expect(wallet.get(0)).toEqual({ address: 'my_address' });
		});

		it('should add account with lowercase address', () => {
			const result = wallet.add({ address: 'myAddress' } as never);

			expect(result).toBeTruthy();
			expect(accountProvider.privateKeyToAccount).toHaveBeenCalledTimes(0);
			expect(wallet).toHaveLength(1);
			expect(wallet.get('myaddress')).toEqual({ address: 'myAddress' });
		});

		it('should override account object for existing address', () => {
			const account1 = { address: 'address', privateKey: 'pkey1' } as never;
			const account2 = { address: 'address', privateKey: 'pkey2' } as never;

			wallet.add(account1);
			expect(wallet.get('address')).toEqual(account1);

			wallet.add(account2);
			expect(wallet.get('address')).toEqual(account2);
		});
	});

	describe('get', () => {
		it('should get account for given index', () => {
			const account = { address: 'my_address' } as never;

			wallet.add(account);

			expect(wallet.get(0)).toEqual(account);
		});

		it('should get account for given address', () => {
			const account = { address: 'my_address' } as never;

			wallet.add(account);

			expect(wallet.get('my_address')).toEqual(account);
		});

		it('should get account for given address only in lower case', () => {
			const account = { address: 'my_Address' } as never;

			wallet.add(account);

			expect(wallet.get('my_address')).toEqual(account);
			expect(wallet.get('my_Address')).toBeUndefined();
		});
	});

	describe('remove', () => {
		it('should remove account for given index', () => {
			const account = { address: 'my_address' } as never;
			wallet.add(account);
			expect(wallet).toHaveLength(1);

			const result = wallet.remove(0);

			expect(result).toBeTruthy();
			expect(wallet).toHaveLength(0);
			expect(wallet.get(0)).toBeUndefined();
		});

		it('should return false if index not found', () => {
			const account = { address: 'my_address' } as never;
			wallet.add(account);
			expect(wallet).toHaveLength(1);

			const result = wallet.remove(2);

			expect(result).toBeFalsy();
			expect(wallet).toHaveLength(1);
		});

		it('should remove account for given address', () => {
			const account = { address: 'my_address' } as never;
			wallet.add(account);
			expect(wallet).toHaveLength(1);

			const result = wallet.remove('my_address');

			expect(result).toBeTruthy();
			expect(wallet).toHaveLength(0);
			expect(wallet.get(0)).toBeUndefined();
		});

		it('should return false if given address not found', () => {
			const account = { address: 'my_address' } as never;
			wallet.add(account);
			expect(wallet).toHaveLength(1);

			const result = wallet.remove('my_address2');

			expect(result).toBeFalsy();
			expect(wallet).toHaveLength(1);
		});
	});

	describe('clear', () => {
		it('should remove all accounts', () => {
			wallet.add({ address: 'my_address' } as never);
			wallet.add({ address: 'my_address2' } as never);
			expect(wallet).toHaveLength(2);

			wallet.clear();

			expect(wallet).toHaveLength(0);
			expect(wallet.get(0)).toBeUndefined();
		});
	});

	describe('encrypt', () => {
		it('should encrypt all accounts and return array', () => {
			const account1 = {
				address: 'my_address1',
				encrypt: jest.fn().mockReturnValue('encrypted_account1'),
			} as any;
			const account2 = {
				address: 'my_address2',
				encrypt: jest.fn().mockReturnValue('encrypted_account2'),
			} as any;
			const options = { myOptions: 'myOptions' };
			wallet.add(account1);
			wallet.add(account2);

			const result = wallet.encrypt('password', options);

			expect(account1.encrypt).toHaveBeenCalledTimes(1);
			expect(account1.encrypt).toHaveBeenCalledWith('password', options);
			expect(account2.encrypt).toHaveBeenCalledTimes(1);
			expect(account2.encrypt).toHaveBeenCalledWith('password', options);
			expect(result).toEqual(['encrypted_account1', 'encrypted_account2']);
		});
	});

	describe('decrypt', () => {
		it('should decrypt all accounts and add to wallet', () => {
			const encryptedAccount1 = 'encrypted_account1';
			const encryptedAccount2 = 'encrypted_account2';
			const account1 = { address: 'my_address1' } as any;
			const account2 = { address: 'my_address2' } as any;
			const options = { myOptions: 'myOptions' };

			when(accountProvider.decrypt)
				.calledWith(encryptedAccount1, 'password', options)
				.mockReturnValue(account1);
			when(accountProvider.decrypt)
				.calledWith(encryptedAccount2, 'password', options)
				.mockReturnValue(account2);

			wallet.decrypt([encryptedAccount1, encryptedAccount2], 'password', options);

			expect(accountProvider.decrypt).toHaveBeenCalledTimes(2);
			expect(accountProvider.decrypt).toHaveBeenCalledWith(
				encryptedAccount1,
				'password',
				options,
			);
			expect(accountProvider.decrypt).toHaveBeenCalledWith(
				encryptedAccount2,
				'password',
				options,
			);
			expect(wallet).toHaveLength(2);
			expect(wallet.get(0)).toEqual(account1);
			expect(wallet.get(1)).toEqual(account2);
		});
	});

	describe('save', () => {
		it('should throw error if local storage not present', () => {
			jest.spyOn(Wallet, 'getStorage').mockReturnValue(undefined);

			expect(() => wallet.save('password')).toThrow('Local storage not available.');
		});

		it('should encrypt wallet and store with local storage for given key', () => {
			const encryptedWallet = ['encryptedWallet'];
			jest.spyOn(wallet, 'encrypt').mockReturnValue(encryptedWallet);

			wallet.save('password', 'myKey');

			expect(wallet.encrypt).toHaveBeenCalledTimes(1);
			expect(wallet.encrypt).toHaveBeenCalledWith('password');
			expect(localStorageSpy.setItem).toHaveBeenCalledTimes(1);
			expect(localStorageSpy.setItem).toHaveBeenCalledWith(
				'myKey',
				JSON.stringify(encryptedWallet),
			);
		});

		it('should encrypt wallet and store with local storage with default key', () => {
			const encryptedWallet = ['encryptedWallet'];
			jest.spyOn(wallet, 'encrypt').mockReturnValue(encryptedWallet);

			wallet.save('password');

			expect(wallet.encrypt).toHaveBeenCalledTimes(1);
			expect(wallet.encrypt).toHaveBeenCalledWith('password');
			expect(localStorageSpy.setItem).toHaveBeenCalledTimes(1);
			expect(localStorageSpy.setItem).toHaveBeenCalledWith(
				'web3js_wallet',
				JSON.stringify(encryptedWallet),
			);
		});
	});

	describe('load', () => {
		it('should throw error if local storage not present', () => {
			jest.spyOn(Wallet, 'getStorage').mockReturnValue(undefined);

			expect(() => wallet.load('password')).toThrow('Local storage not available.');
		});

		it('should load wallet from local storage for given key and decrypt', () => {
			const encryptedWallet = JSON.stringify(['encryptedWallet']);

			when(localStorageSpy.getItem).calledWith('myKey').mockReturnValue(encryptedWallet);
			jest.spyOn(wallet, 'decrypt').mockReturnValue({} as never);

			wallet.load('password', 'myKey');

			expect(wallet.decrypt).toHaveBeenCalledTimes(1);
			expect(wallet.decrypt).toHaveBeenCalledWith(['encryptedWallet'], 'password');
			expect(localStorageSpy.getItem).toHaveBeenCalledTimes(1);
			expect(localStorageSpy.getItem).toHaveBeenCalledWith('myKey');
		});

		it('should load wallet from local storage for default key and decrypt', () => {
			const encryptedWallet = JSON.stringify(['encryptedWallet']);

			when(localStorageSpy.getItem)
				.calledWith('web3js_wallet')
				.mockReturnValue(encryptedWallet);
			jest.spyOn(wallet, 'decrypt').mockReturnValue({} as never);

			wallet.load('password');

			expect(wallet.decrypt).toHaveBeenCalledTimes(1);
			expect(wallet.decrypt).toHaveBeenCalledWith(['encryptedWallet'], 'password');
			expect(localStorageSpy.getItem).toHaveBeenCalledTimes(1);
			expect(localStorageSpy.getItem).toHaveBeenCalledWith('web3js_wallet');
		});
	});
});
