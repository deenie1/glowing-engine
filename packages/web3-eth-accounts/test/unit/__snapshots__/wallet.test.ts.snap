// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet create should create given # of accounts using the account provider 1`] = `
[
  {
    "address": "account_create_1",
  },
  {
    "address": "account_create_2",
  },
  {
    "address": "account_create_3",
  },
  {
    "address": "account_create_4",
  },
  {
    "address": "account_create_5",
  },
  {
    "address": "account_create_6",
  },
  {
    "address": "account_create_7",
  },
  {
    "address": "account_create_8",
  },
  {
    "address": "account_create_9",
  },
  {
    "address": "account_create_10",
  },
]
`;
